yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+   * by the timeline service
+   * @throws SparkException always
+   */
+  override def writeEventLogs(appId: String, attemptId: Option[String],
+      zipStream: ZipOutputStream): Unit = {
+    throw new SparkException("Unsupported Feature")
+  }
+
+  override def toString(): String = {
+    s"YarnHistoryProvider bound to history server at $timelineEndpoint," +
+    s" enabled = $enabled;" +
+    s" refresh count = ${getRefreshCount()}; failed count = ${getRefreshFailedCount()};" +
+    s" last update ${applications.updated};" +
+    s" history size ${applications.size};" +
+    s" ${refresher}"
+  }
+
+  /**
+   * Comparison function that defines the sort order for the application listing.
+   *
+   * @return Whether `i1` should precede `i2`.
+   */
+  private def compareAppInfo(
+      i1: ApplicationHistoryInfo,
+      i2: ApplicationHistoryInfo): Boolean = {
+    val a1 = i1.attempts.head
+    val a2 = i2.attempts.head
+    if (a1.endTime != a2.endTime) a1.endTime >= a2.endTime else a1.startTime >= a2.startTime
+  }
+
+
+  /**
+   * This is the implementation of the triggered refresh logic.
+   * It awaits events
+   */
+
+  private[spark] class Refresher extends Runnable {
+
+    sealed trait RefreshActions;
+    /** start the refresh **/
+    case class Start() extends RefreshActions;
+    /** refresh requested at the given time */
+    case class RefreshRequest(time: Long) extends RefreshActions;
+    /** stop */
+    case class StopExecution() extends RefreshActions;
+
+    private val queue = new LinkedBlockingQueue[RefreshActions]()
+    private val running = new AtomicBoolean(false)
+    private var self: Thread = _
+    private val _lastRefreshAttemptTime = new AtomicLong(0)
+    private val _messagesProcessed = new AtomicLong(0)
+    private val _refreshesExecuted = new AtomicLong(0)
+
+    /**
+     * Bond to the thread then start it
+     * @param t thread
+     */
+    def start(t: Thread) {
+      this.synchronized {
+        self = t;
+        running.set(true)
+        queue.add(Start())
+        t.start()
+      }
+    }
+
+    /**
+     * Request a refresh. If the request queue is empty, a refresh request
+     * is queued.
+     * @param time time request was made
+     */
+    def refresh(time: Long): Unit = {
+      if (queue.isEmpty) {
+        queue.add(RefreshRequest(time))
+      }
+    }
+
+    /**
+     * Stop operation.
+     * @return true if the stop was scheduled
+     */
+    def stopRefresher(): Boolean = {
+      this.synchronized {
+        if (isRunning()) {
+          // yes, more than one stop may get issued. but it will
+          // replace the previous one.
+          queue.clear()
+          queue.add(StopExecution())
+          self.interrupt()
+          true
+        } else {
+          false
+        }
+      }
+    }
+
+    /**
+     * Thread routine
+     */
+    override def run(): Unit = {
+      try {
+        var stopped = false;
+        while (!stopped) {
+          take match {
```

Comment:
I'd have agreed with you about the over-engineering until about 10 days ago. It's need becomes apparent once you have a few hundred applications in the history. While the windowed GET reduces the amount of data to be marshalled, you are still making regular requests and asking for the JSON back. And, as those updates need to include all incomplete applications (how else do you know they've finished), old-yet-incomplete applications can lead to large windows.  The implementation in this patch doesn't have any scheduled load at all, except at startup.

I'm not a great fan of the UX (the refresh is triggered by the web UI but isn't immediate), but don't have a good solution for scale on large YARN clusters without going near the web UI & again, adding a refresh meta-header if hinted by the history provider.